# Lab 01: Foundations of AI Engineering & Project Setup

## Task 1: Repository Directory Structure

In [1]:
import os

folders = ['data/raw', 'data/processed', 'models', 'notebooks', 'src', 'configs', 'logs']

for folder in folders:
    os.makedirs(folder, exist_ok=True)
    open(folder + '/.gitkeep', 'w').close()

In [2]:
%%writefile .gitignore
__pycache__/
*.py[cod]
.ipynb_checkpoints/
.venv/
aipdd_env/
venv/
env/
*.pt
*.onnx
*.h5
.DS_Store

Overwriting .gitignore


In [3]:
for root, dirs, files in os.walk('.'):
    dirs[:] = [d for d in dirs if d not in ['.git', '.ipynb_checkpoints', 'aipdd_env', '.venv']]
    level = root.count(os.sep)
    print('    ' * level + os.path.basename(os.path.abspath(root)) + '/')
    for file in files:
        print('    ' * (level + 1) + file)

AI_Project/
    .gitignore
    Lab01_Project_Setup.ipynb
    lab01_terminal_commands.txt
    configs/
        .gitkeep
    data/
        processed/
            .gitkeep
        raw/
            .gitkeep
    logs/
        .gitkeep
    models/
        .gitkeep
    notebooks/
        .gitkeep
    src/
        .gitkeep


## Task 2: Virtual Environment and requirements.txt

In [4]:
import sys

!"{sys.executable}" -m pip install numpy opencv-python matplotlib ultralytics jupyter


[notice] A new release of pip is available: 24.0 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [5]:
import sys

!"{sys.executable}" -m pip freeze > requirements.txt

In [6]:
print(open('requirements.txt').read())

anyio==4.15.1
argon2-cffi==25.1.0
argon2-cffi-bindings==26.1.0
arrow==1.4.0
asttokens==3.0.2
async-lru==2.3.0
attrs==26.1.0
babel==2.18.0
beautifulsoup4==4.15.0
bleach==6.4.0
certifi==2026.7.22
cffi==2.1.1
charset-normalizer==3.5.2
colorama==0.4.6
comm==0.2.3
contourpy==1.3.3
cycler==0.12.1
debugpy==1.8.22
defusedxml==0.7.1
executing==2.2.1
fastjsonschema==2.22.2
filelock==4.0.10
fonttools==4.66.1
fqdn==1.6.0
fsspec==2026.9.0
h11==0.16.0
httpcore==1.0.9
httpx==0.28.1
idna==3.20
ipykernel==7.4.0
ipython==9.17.1
ipython_pygments_lexers==1.1.1
ipywidgets==8.1.9
isoduration==20.11.0
jedi==0.20.0
Jinja2==3.1.6
json5==0.15.0
jsonpointer==3.1.1
jsonschema==4.26.0
jsonschema-specifications==2025.9.1
jupyter==1.1.1
jupyter-console==6.6.3
jupyter-events==0.12.1
jupyter-lsp==2.3.1
jupyter_builder==1.2.3
jupyter_client==8.10.0
jupyter_core==5.9.1
jupyter_server==2.21.1
jupyter_server_terminals==0.5.4
jupyterlab==4.6.4
jupyterlab_pygments==0.3.0
jupyterlab_server==2.28.1
jupyterlab_widgets==3.0.17


## Task 3: Git Initialization and Branching

In [7]:
!git init
!git branch -M main
!git config user.name "Abdul Rehman Ali"
!git config user.email "your_email@example.com"
!git add .gitignore data models notebooks src configs logs
!git commit -m "Initial commit: project structure and gitignore"

Initialized empty Git repository in C:/Users/Abdul Rehman Ali/OneDrive/Documents/vs code/AIPDD/Lab01/AI_Project/.git/
[main (root-commit) 2a79b4a] Initial commit: project structure and gitignore
 8 files changed, 11 insertions(+)
 create mode 100644 .gitignore
 create mode 100644 configs/.gitkeep
 create mode 100644 data/processed/.gitkeep
 create mode 100644 data/raw/.gitkeep
 create mode 100644 logs/.gitkeep
 create mode 100644 models/.gitkeep
 create mode 100644 notebooks/.gitkeep
 create mode 100644 src/.gitkeep


In [8]:
!git checkout -b feature/lab01-setup
!git add requirements.txt
!git commit -m "Add requirements.txt"
!git checkout main
!git merge --no-ff feature/lab01-setup -m "Merge feature/lab01-setup into main"

Switched to a new branch 'feature/lab01-setup'


[feature/lab01-setup c954d6b] Add requirements.txt
 1 file changed, 119 insertions(+)
 create mode 100644 requirements.txt


Switched to branch 'main'


Merge made by the 'ort' strategy.
 requirements.txt | 119 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 119 insertions(+)
 create mode 100644 requirements.txt


In [9]:
!git status
!git log --oneline --graph --all
!git branch

On branch main
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	Lab01_Project_Setup.ipynb
	lab01_terminal_commands.txt

nothing added to commit but untracked files present (use "git add" to track)
*   b6d755f Merge feature/lab01-setup into main
|\  
| * c954d6b Add requirements.txt
|/  
* 2a79b4a Initial commit: project structure and gitignore
  feature/lab01-setup
* main


## Task 4: Jupyter Kernel and Package Versions

In [10]:
import sys
print(sys.executable)

c:\Users\Abdul Rehman Ali\OneDrive\Documents\vs code\AIPDD\aipdd_env\Scripts\python.exe


In [11]:
import platform
import numpy
import cv2
import torch
import ultralytics
import matplotlib

print('Python', platform.python_version())
print('NumPy', numpy.__version__)
print('OpenCV', cv2.__version__)
print('PyTorch', torch.__version__)
print('Ultralytics', ultralytics.__version__)
print('Matplotlib', matplotlib.__version__)

Creating new Ultralytics Settings v0.0.8 file  
View Ultralytics Settings with 'yolo settings' or at 'C:\Users\Abdul Rehman Ali\AppData\Roaming\Ultralytics\settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/usage/settings.
Python 3.11.9
NumPy 2.4.6
OpenCV 5.0.0
PyTorch 2.14.1+cpu
Ultralytics 8.4.173
Matplotlib 3.11.2


## Task 5: Project Specification

# Smart Surveillance and Object Detection System

## Problem Statement
Watching camera feeds by hand is slow and events get missed. This project detects objects in images and live video and records important events automatically.

## Core Objectives
* **Detect** people, vehicles and other objects with YOLO
* **Preprocess** images with OpenCV for consistent model input
* **Alert** and log security events with timestamps
* **Organize** the work in a clean Git repository

### Environment
* Python 3.10 or newer
* Virtual environment: `aipdd_env`
* Libraries: NumPy, OpenCV, Matplotlib, Ultralytics, Jupyter
```